File: website/src/__generated/services/ForumController.ts
import type { Executor } from '../';
import type { ForumDto } from '../model/dto';
import type { Page } from '../model/static';

export class ForumController {
    
    constructor(private executor: Executor) {}
    
    async findForum(options: ForumControllerOptions['findForum']): Promise<
        ForumDto['ForumController/DEFAULT_FORUM'] | undefined
    > {
        let _uri = '/categories/forums/';
        _uri += encodeURIComponent(options.id);
        _uri += '/';
        return (await this.executor({uri: _uri, method: 'GET'})) as ForumDto['ForumController/DEFAULT_FORUM'] | undefined
    }
    
    async findForums(options: ForumControllerOptions['findForums']): Promise<
        Page<ForumDto['ForumController/DEFAULT_FORUM']>
    > {
        let _uri = '/categories/';
        _uri += encodeURIComponent(options.categoryId);
        _uri += '/forums/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.page;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'page='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        _value = options.size;
        if (_value !== undefined && _value !== null) {
            _uri += _separator
            _uri += 'size='
            _uri += encodeURIComponent(_value);
            _separator = '&';
        }
        return (await this.executor({uri: _uri, method: 'GET'})) as Page<ForumDto['ForumController/DEFAULT_FORUM']>
    }
}

export type ForumControllerOptions = {
    'findForum': {readonly id: string},
    'findForums': {
        readonly categoryId: string, 
        readonly page?: number, 
        readonly size?: number
    }
}

File: server/src/main/kotlin/cn/enaium/zensquare/controller/category/forum/ForumController.kt
/*
 * ZenSquare is an opensource forums
 *
 * Copyright (C) 2023  Enaium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package cn.enaium.zensquare.controller.category.forum

import cn.enaium.zensquare.model.entity.Forum
import cn.enaium.zensquare.model.entity.by
import cn.enaium.zensquare.repository.ForumRepository
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.util.*

/**
 * @author Enaium
 */
@RestController
class ForumController(val forumRepository: ForumRepository) {
    @GetMapping("/categories/forums/{id}")
    fun findForum(@PathVariable id: UUID) = forumRepository.findById(id)

    @GetMapping("/categories/{categoryId}/forums")
    fun findForums(
        @PathVariable categoryId: UUID,
        @RequestParam(defaultValue = "0") page: Int = 0,
        @RequestParam(defaultValue = "10") size: Int = 10
    ): Page<@FetchBy("DEFAULT_FORUM") Forum> {
        return forumRepository.findAllByCategoryId(categoryId, DEFAULT_FORUM, PageRequest.of(page, size))
    }

    companion object {
        val DEFAULT_FORUM = newFetcher(Forum::class).by {
            allScalarFields()
            thread()
        }
    }
}